lib/core/src/Cardano/Wallet/Primitive/BlockSummary.hs

```diff
@@ -0,0 +1,200 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+-- |
+-- Copyright: © 2022 IOHK
+-- License: Apache-2.0
+--
+-- This module provides the 'BlockSummary' type
+-- which summarizes a contiguous sequence of blocks.
+--
+module Cardano.Wallet.Primitive.BlockSummary
+    ( BlockSummary (..)
+    , LightSummary
+
+    -- * Chain Events
+    , ChainEvents
+    , fromBlockEvents
+    , toAscBlockEvents
+
+    -- * Block Events
+    , BlockEvents (..)
+    , fromEntireBlock
+
+    -- * Testing
+    , summarizeOnTxOut
+    , mkChainEvents
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( RewardAccount )
+import Cardano.Wallet.Primitive.Types
+    ( Block (..)
+    , BlockHeader (..)
+    , DelegationCertificate
+    , Slot
+    , chainPointFromBlockHeader
+    , dlgCertAccount
+    , toSlot
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( Tx (..), TxOut (..) )
+import Data.Functor.Identity
+    ( Identity (..) )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Map
+    ( Map )
+import Data.Quantity
+    ( Quantity )
+import Data.Word
+    ( Word32 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Types as Block
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+    BlockSummary
+-------------------------------------------------------------------------------}
+-- | A 'BlockSummary' summarizes the data contained in a contiguous sequence
+-- of blocks.
+--
+-- However, instead of storing the sequence of blocks of directly as a Haskell
+-- list, the 'BlockSummary' only provides a 'query' function
+-- which looks up all transactions associated to a given addresses.
```

Comment:
in next installements we will extend that and allow discovery of other things, like stake keys, policy keys, certificates, etc. Right?

Reply:
All things that we need in order to do address discovery, indeed.

The type `BlockSummary` is actually polymorphic in the `addr` type, and I felt that the gist would be easier to understand if I kept calling it "address".

In the `LightSummary` type, we currently instantiate this to `addr = Either Address RewardAccount` — as this is all the information currently required by `applyBlocks`.